pkg/cluster/loadbalancerprofile.go

```diff
@@ -321,27 +317,51 @@ func getDesiredOutboundIPs(managedOBIPCount int, ipAddresses map[string]mgmtnetw
 }
 
 func (m *manager) createPublicIPAddresses(ctx context.Context, ipAddresses map[string]mgmtnetwork.PublicIPAddress, numToCreate int) error {
-	ch := make(chan createIPResult)
-	defer close(ch)
-	var errResults []string
-	// create additional IPs if needed
-	for i := 0; i < numToCreate; i++ {
-		go m.createPublicIPAddress(ctx, ch)
-	}
+	var wg sync.WaitGroup
+
+	resultCh := make(chan mgmtnetwork.PublicIPAddress)
+	errCh := make(chan error)
+	doneCh := make(chan bool)
 
+	m.processIPAddresses(ctx, numToCreate, resultCh, errCh, &wg)
+
+	go func() {
+		wg.Wait()
+		close(resultCh)
+		close(errCh)
+		doneCh <- true
+		close(doneCh)
+	}()
+
+	return consumResults(resultCh, errCh, doneCh, ipAddresses)
+}
+
+func (m *manager) processIPAddresses(ctx context.Context, numToCreate int, resultCh chan<- mgmtnetwork.PublicIPAddress, errCh chan<- error, wg *sync.WaitGroup) {
+	wg.Add(numToCreate)
 	for i := 0; i < numToCreate; i++ {
-		result := <-ch
-		if result.err != nil {
-			errResults = append(errResults, fmt.Sprintf("creation of ip address %s failed with error: %s", *result.ip.Name, result.err.Error()))
-		} else {
-			ipAddresses[*result.ip.Name] = result.ip
-		}
+		go m.createPublicIPAddress(ctx, resultCh, errCh, wg)
 	}
+}
 
-	if len(errResults) > 0 {
-		return fmt.Errorf("failed to create required IPs\n%s", strings.Join(errResults, "\n"))
+func consumResults(resultCh <-chan mgmtnetwork.PublicIPAddress, errCh <-chan error, doneCh <-chan bool, ipAddresses map[string]mgmtnetwork.PublicIPAddress) error {
+	var errResults []string
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				errResults = append(errResults, err.Error())
+			}
```

Comment:
we could do 
```go
case err, ok := <-errCh:
	if ok {
		errResults = append(errResults, err.Error())
	}
```
instead, but checking against nil (the zeroed value of error) is more straightforward and easy to read.

Same for the "ip := <-resultCh:" case.

Reply:
If we defer closing `errCh` and `resultCh`, can we remove nil check, because it won't read the zero value from the channels anymore?